### PR TITLE
Generic Hyperparameter Seacher 

### PR DESCRIPTION
**Pull request by CaitiCellier**
Initial commit for a hyperparameter searcher using sampling
Includes classes for dynamic parameters that are used for tuning

### CODE REVIEW

File: src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import com.etsy.conjecture.scalding.train._
+import com.twitter.scalding.Dsl._
+import com.twitter.scalding._
+
+import scala.util.Random
+
+/**
+  * Samples random parameter values to perform a fast efficient hyperparameter search
+  */
+abstract class HyperparameterSearcher[L <: Label, M <: UpdateableModel[L, M], E <: ModelEvaluation[L]]
```
Comment (jattenberg) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: type wizardry.

Comment (CaitiCellier) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: Yea sorry about that haha I wanted it to be able to be used for all types of labels and models.  When I open the review branch for the BigData side it'll be easier to see how it comes into play.

File: src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    instances
+      .flatMapTo('instance -> ('instance, 'trial)) {
+         instance: LabeledInstance[L] =>
+             settings.keySet.map {
+               trial: Int =>
+                  (instance, trial)
+             }
+       }
+       .groupBy('trial) {
+          _.toList[LabeledInstance[L]]('instance -> 'instances).reducers(1000)
```
Comment (rjhall) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: This is ok for now but going forward the parallelization should be up to the user with some default, instead of hard coded.

**Comment by jattenberg** [comment]
I think this might benefit from a tiny example job

File: src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+class MulticlassHyperparameterSearcher(option: DynamicOptions, parameters: Seq[HyperParameter[_]], numTrials: Int, categories: Array[String]) extends HyperparameterSearcher[MulticlassLabel, UpdateableMulticlassLinearModel, MulticlassModelEvaluation](option, parameters, numTrials) {
+    val evaluator = new MulticlassEvaluator(categories)
+    def getModelTrainer(args: Args) = new MulticlassModelTrainer(args, categories)
+}
+
+trait ParameterSampler[T] extends Serializable {
+  def sample(rng: scala.util.Random): T
+  def buckets: Array[(T,Double,Double,Double,Int)]
```
Comment (rjhall) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: can you put in some comments aout what this is doing and what the stuff in the array is

File: src/main/scala/com/etsy/conjecture/scalding/util/DynamicOptions.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
+package com.etsy.conjecture.scalding.util
+
```
Comment (rjhall) [inline src/main/scala/com/etsy/conjecture/scalding/util/DynamicOptions.scala]: whats the point of using this instead of Args everywhere

Comment (CaitiCellier) [inline src/main/scala/com/etsy/conjecture/scalding/util/DynamicOptions.scala]: I thought if I used Args everywhere I would have to do type checks and conversations everywhere and it might be messier?  I wanted a something to hold the Args in their actual types that I could also maybe in the future add more operations too.  I'm also just used to having an options class inside of a toolkit, seemed like a better long term solution for more complex things in the future.  But that might be a preference thing.

File: src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala (hunk 2 above)
Comment (CaitiCellier) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: You mean the reducer amount?

Comment (jattenberg) [inline src/main/scala/com/etsy/conjecture/scalding/util/HyperparameterSearcher.scala]: yes

**Comment by CaitiCellier** [comment]
There's an interface class which must belong in BigData, relatively short but just so you can see all components here's a gist of it:
https://github.etsycorp.com/gist/anonymous/93c3faaf809dc677082c
